File: 2024/12/main.kt
import java.io.File
import kotlin.math.abs
import kotlin.time.measureTime

fun main() {
    val inputTest = getInput("inputTest.txt")
    val input = getInput("input.txt")
    println("-------- PART 1 --------")
    println(inputTest)
    println(input.lines.first())
    println(input.lines.last())
    //mainMeasuringTime({ part1(inputTest) }, "InputTest")
    //mainMeasuringTime({ part1(input) }, "Input")
    println("\n-------- PART 2 --------")
    mainMeasuringTime({ part2(inputTest) }, "InputTest")
    //mainMeasuringTime({ part2(input) }, "Input")
}

fun mainMeasuringTime(someFun: () -> Long, toPrint: String) {
    val funTime = measureTime {
        val result = someFun()
        println("$toPrint $result")
    }
    println("$toPrint exec time $funTime")

}

private fun getInput(fileName: String): Input {
    val file = File(fileName).readText()
    val lines = file.split("\n").dropLast(1)

    val maxY = lines.size - 1
    val maxX = lines[0].length - 1
    return Input(lines, maxX, maxY)
}

data class Input(val lines: List<String>, val maxX: Int, val maxY: Int)
data class Pose(val x: Int, val y: Int)
data class Group(val char: Char, val poses: List<Pose>)

private fun part2(input: Input): Long {
    val groups = input.lines.withIndex().fold(listOf<Pair<Char, List<Group>>>()) { acc, (y, value) ->
        value.withIndex().fold(acc) { charToGroupsMap, (x, char) ->
            val pose = Pose(x, y)
            val currentGroups = charToGroupsMap.filter { it.first == char }.map { it.second }.flatten()

            if (pose in currentGroups.flatMap { it.poses }) {
                charToGroupsMap
            } else {

                val group = exploreGroup(char, listOf(pose), input, Group(char, listOf(pose)), setOf())
                val groups = currentGroups.plus(group)
                println("New group done ${group.char}, ${group.poses.size} Groups count ${groups.size}")
                val pair = char to listOf(group)
                charToGroupsMap.plus(pair)
            }
        }
    }.map { it.second }.flatten().distinct()
    println("Groups done")
    println(groups.joinToString("\n"))
    return groups.sumOf { group: Group ->
        val borders = group.poses.map { pose ->
            val next = nextPoses(pose)
            next.filter { p -> input.lines.getOrNull(p.y)?.getOrNull(p.x) != group.char }
        }.distinct().flatten()
        val vertical = borders.sortedBy { it.x }.sortedBy { it.y }.groupBy { it.x }.filter { it.value.size > 1 }
        val okokV = vertical.values.fold(0) { acc, poses: List<Pose> ->
            acc + poses.withIndex().fold(0) { poseAcc, (i, p) ->
                val prevOrNull = poses.getOrNull(i - 1)
                if (prevOrNull == null || abs(prevOrNull.x - p.x) > 1) {
                    acc + 1
                } else acc

            }
        }


        val horizontal = borders.sortedBy { it.y }.sortedBy { it.x }.groupBy { it.y }
        val okokH = horizontal.values.fold(0) { acc, poses: List<Pose> ->
            acc + poses.withIndex().fold(0) { poseAcc, (i, p) ->
                val prevOrNull = poses.getOrNull(i - 1)
                if (prevOrNull == null || abs(prevOrNull.x - p.x) > 1) {
                    val char = group.char
                    println("found vertice at $p for $char")
                    acc + 1
                } else acc

            }
        }

        println("Vertices V:$okokV, H:$okokH")
        //println(vertices.joinToString("\n"))
        val perimeter = okokV + okokH
        val area = group.poses.size
        val result = perimeter * area
        println("A region of ${group.char} plants with price $area * $perimeter = $result.")
        result
    }.toLong()
}

private fun part1(input: Input): Long {
    val groups = input.lines.withIndex().fold(listOf<Pair<Char, List<Group>>>()) { acc, (y, value) ->
        value.withIndex().fold(acc) { charToGroupsMap, (x, char) ->
            val pose = Pose(x, y)
            val currentGroups = charToGroupsMap.filter { it.first == char }.map { it.second }.flatten()

            if (pose in currentGroups.flatMap { it.poses }) {
                charToGroupsMap
            } else {

                val group = exploreGroup(char, listOf(pose), input, Group(char, listOf(pose)), setOf())
                val groups = currentGroups.plus(group)
                println("New group done ${group.char}, ${group.poses.size} Groups count ${groups.size}")
                val pair = char to listOf(group)
                charToGroupsMap.plus(pair)
            }
        }
    }.map { it.second }.flatten().distinct()
    println("Groups done")
    println(groups.joinToString("\n"))
    return groups.sumOf { group: Group ->
        val borders = group.poses.flatMap { pose ->
            val next = nextPoses(pose)
            next.filter { p -> input.lines.getOrNull(p.y)?.getOrNull(p.x) != group.char }
        }
        println("borders")
        println(borders)
        val perimeter = borders.size //.filter { it == null }.size +  groupsOrNull.filterNotNull().sumOf { it.area() }
        val area = group.poses.size
        val result = perimeter * area
        println("A region of ${group.char} plants with price $area * $perimeter = $result.")
        result
    }.toLong()
}

tailrec fun exploreGroup(char: Char, poses: List<Pose>, input: Input, group: Group, seen: Set<Pose>): Group {
    println("Poses len ${poses.size}")
    val next = nextPosesInGroup(char, poses, input).filter { p -> p !in group.poses }
    val newGroup = group.copy(poses = group.poses + next)
    if (next.isEmpty()) {
        return newGroup
    }
    val filtered = next.filter {
        it !in seen
    }
    return exploreGroup(char, filtered, input, newGroup, seen + poses)
}

fun nextPosesInGroup(char: Char, poses: List<Pose>, input: Input): List<Pose> {
    return listOf((1 to 0), (0 to 1), (-1 to 0), (0 to -1)).map { (dx, dy) ->
        poses.map { pose ->
            Pose(pose.x + dx, pose.y + dy)
        }
    }.flatten().filter { p ->
        (p.x <= input.maxX && p.y <= input.maxY && p.x >= 0 && p.y >= 0)
                &&
                char == input.lines[p.y][p.x]
    }.distinct()
}

fun nextPoses(pose: Pose): List<Pose> {
    return listOf((1 to 0), (0 to 1), (-1 to 0), (0 to -1)).map { (dx, dy) ->
        Pose(pose.x + dx, pose.y + dy)
    }
}

File: 2024/12/main.go
package main

import (
	//"github.com/samber/lo"
	_ "embed"
	"fmt"
	"math"
	"os"
	"reflect"
	"strconv"
	"strings"
	"time"

	//"github.com/samber/lo"
)

type Input struct {
	lines []string
	maxX  int
	maxY  int
}

func parse(value string) Input {
	lines := strings.Split(value, "\n")
	maxY := len(lines) - 2
	maxX := len(lines[0]) - 1
	return Input{
		lines,
		maxX,
		maxY,
	}
}

type search struct {
	area      int
	perimeter int
}
type move struct {
	dx int
	dy int
}

type Pose struct {
	x int
	y int
}

type Group struct {
	poses []Pose
}


func part1(input Input) int {
	posesMap := map[string][]Pose{}
	for y, line := range input.lines {
		for x, char := range line {
			oldPoses := posesMap[string(char)]
			posesMap[string(char)] = append(oldPoses, Pose{x, y})
		}
	}

	//groups := map[string][]Group
	currentPoses := []Pose{{0, 0}}
	condition := true
	for condition {
		fmt.Printf("Current poses %v", currentPoses)
		fmt.Println()
		nextPoses := nextPoses(currentPoses, input)

		if reflect.DeepEqual(nextPoses, currentPoses) {
			condition = false
			return 42
		}
		currentPoses = nextPoses
	} 
	return 1
}

func nextPoses(poses []Pose, input Input) []Pose{
	newPoses := []Pose{}
	for _, pose := range poses {
		nextRight := Pose{pose.x + 1, pose.y}
		if (nextRight.x <= input.maxX && nextRight.y <= input.maxY) {
			newPoses = append(newPoses, nextRight)
		}

		nextBottom := Pose{pose.x, pose.y + 1}
		if (nextBottom.x <= input.maxX && nextBottom.y <= input.maxY) {
			newPoses = append(newPoses, nextBottom)
		}
	}
	return newPoses
}

func areAdjacent(pose Pose, otherPose Pose) bool {
	if math.Abs(float64(pose.x-otherPose.x)) <= 1 && math.Abs(float64(otherPose.y-pose.y)) <= 1 {
		return true
	}
	return false
}

func part2(input Input) int {
	return 2
}

//go:embed inputTest.txt
var inputTest string

//go:embed input.txt
var input string

func main() {
	part, err := strconv.Atoi(os.Args[1])
	if err != nil {
		panic("Part 1 or part 2?")
	}

	runInput := inputTest
	inputName := "inputTest"
	if len(os.Args) > 2 && os.Args[2] == "go" {
		inputName = "input"
		runInput = input
	}
	fmt.Printf("Part %v %v", part, inputName)
	run(runInput, part)
}

func timeTrack(start time.Time) {
	elapsed := time.Since(start)
	fmt.Printf("Took %v", elapsed)
}

func run(input string, part int) {
	defer timeTrack(time.Now())

	var result int

	if part == 1 {
		result = part1(parse(input))
	} else {
		result = part2(parse(input))
	}
	fmt.Println()
	fmt.Println(result)
}
